### PR TITLE
Disallow foreign key constraint on ddl

### PR DESCRIPTION
**Pull request by harshit-gangal**
## Description

Vitess today allowed foreign keys constraint to be sent down to the mysql to be applied on the table. 

In a sharded keyspace, they might not be correctly set and would fail user queries. 

This change is to provide a `-foreign_key` flag on VTGate with default "allow" but with an option to change it to "disallow" 

## Checklist
- [X] Tests were added or are not required
- [ ] Documentation was added or is not required

## Impacted Areas in Vitess
Components that this PR will affect:

- [X]  Query Serving

### CODE REVIEW

File: go/vt/vtgate/vtgate.go [hunk 1]
```diff
@@ -82,6 +82,8 @@ var (
 	// lockHeartbeatTime is used to set the next heartbeat time.
 	lockHeartbeatTime = flag.Duration("lock_heartbeat_time", 5*time.Second, "If there is lock function used. This will keep the lock connection active by using this heartbeat")
 	warnShardedOnly   = flag.Bool("warn_sharded_only", false, "If any features that are only available in unsharded mode are used, query execution warnings will be added to the session")
+
+	foreignKey = flag.String("foreign_key", "allow", "This is to provide how to handle foreign key constraint in create/alter table. Valid values are: allow, disallow")
```
Comment (deepthi) [inline go/vt/vtgate/vtgate.go:86]: @systay and @shlomi-noach can do the substantive review.
My comment is on the flag name. `foreign_key_mode` may be more descriptive and useful.

**Review by GuptaManan100 — APPROVED** [review]

File: go/vt/vtgate/planbuilder/ddl.go [hunk 2]
```diff
@@ -17,6 +19,33 @@ const (
 	DifferentDestinations string = "Tables or Views specified in the query do not belong to the same destination"
 )
 
+type fkStrategy int
+
+const (
+	allow fkStrategy = iota
+	disallow
```
Comment (shlomi-noach) [inline go/vt/vtgate/planbuilder/ddl.go:26]: The names `allow` and `disallow` seem to be too general for the scope of `planbuilder` package. I suggest renaming as `fkAllow` and `fkDisallow` or similar.

File: go/vt/vtgate/planbuilder/ddl.go [hunk 3]
```diff
@@ -116,6 +141,17 @@ func buildDDLPlans(sql string, ddlStatement sqlparser.DDLStatement, reservedVars
 		}, nil
 }
 
+func checkFKError(vschema ContextVSchema, ddlStatement sqlparser.DDLStatement) error {
+	if fkStrategyMap[strings.ToLower(vschema.ForeignKeyMode())] == disallow {
```
Comment (shlomi-noach) [inline go/vt/vtgate/planbuilder/ddl.go:145]: Seems like every call will have to run `ToLower`. Consider moving `ToLower` into `ForeignKeyMode()` func.

File: go/vt/vtgate/vcursor_impl.go [hunk 4]
```diff
@@ -753,6 +753,11 @@ func (vc *vcursorImpl) WarnUnshardedOnly(format string, params ...interface{}) {
 	}
 }
 
+// ForeignKey implements the VCursor interface
+func (vc *vcursorImpl) ForeignKeyMode() string {
+	return *foreignKeyMode
```
Comment (shlomi-noach) [inline go/vt/vtgate/vcursor_impl.go:758]: consider `return strings.ToLower(*foreignKeyMode)` so that clients can immediately use as map key.

**Review by shlomi-noach — COMMENTED** [review]
Suggesting a few changes, otherwise this looks good!

**Pull request merged**